pallets/fixed-point/src/lib.rs

```diff
@@ -0,0 +1,146 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+//! A pallet that demonstrates the fundamentals of Fixed Point arithmetic.
+//! This pallet implements three multiplicative accumulators using fixed point.
+//!
+//! ## Manual Implementation
+//! Here we use simple u32 values, and just keep in mind that the high-order 16 bits
+//! represent the integer part while the low 16 bits represent fractional places.
+//!
+//! ## Permill Implementation
+//! Here we use Substrate's built-in Permill type. We'll use the saturating_mul function
+//!
+//!
+//! ## Substrate-fixed Implementation
+//! Here we use an external crate called substrate-fixed which implements more advanced
+//! mathematical operations including transcendental functions.
+
+use frame_support::{
+	decl_event,
+	decl_error,
+	decl_module,
+	decl_storage,
+	dispatch::DispatchResult,
+};
+use frame_system::{self as system, ensure_signed};
+use substrate_fixed::types::U16F16;
+use sp_arithmetic::{Permill, traits::Saturating};
+
+#[cfg(not(feature = "std"))]
+use sp_arithmetic::PerThing;
+
+#[cfg(test)]
+mod tests;
+
+pub trait Trait: system::Trait {
+	type Event: From<Event> + Into<<Self as system::Trait>::Event>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Example {
+		/// Permill accumulator, value starts at 1 (multiplicative identity)
+		PermillAccumulator get(fn permill_value): Permill = Permill::one();
+		/// Substrate-fixed accumulator, value starts at 1 (multiplicative identity)
+		FixedAccumulator get(fn fixed_value): U16F16 = U16F16::from_num(1);
+		/// Manual accumulator, value starts at 1 (multiplicative identity)
+		ManualAccumulator get(fn manual_value): u32 = 1 << 16;
+	}
+}
+
+decl_event!(
+	pub enum Event {
+		// For all varients of the event, the contained data is
+		// (new_factor, new_product)
+		/// Permill accumulator has been updated.
+		PermillUpdated(Permill, Permill),
+		/// Substrate-fixed accumulator has been updated.
+		FixedUpdated(U16F16, U16F16),
+		/// Manual accumulator has been updated.
+		ManualUpdated(u32, u32),
+	}
+);
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// Some math operation overflowed
+		Overflow,
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event() = default;
+
+		/// Update the Permill accumulator implementation's value by multiplying it
+		/// by the new factor given in the extrinsic
+		fn update_permill(origin, new_factor: Permill) -> DispatchResult {
+			let _ = ensure_signed(origin)?;
```

Comment:
Here and in the subsequent functions:

```suggestion
			ensure_signed(origin)?;
```

I believe `AccountId` doesn't implement `#[must_use]`.